### DISCUSSION TITLE
Skip verbose output of autocd

### DISCUSSION THREAD

**Discussion by Jai-JAP** [akinomyoga/ble.sh#421] [Ideas]
Currently setting the `autocd` option gives verbose output of the format `cd -- <dir>`

Is it possible to provide a toggle optionally to disable this verbose output? 

There is a hacky fix around this which is as follows
 - `exec {BASH_XTRACEFD}>/dev/null` to disable verbosity
 - `exec {BASH_XTRACEFD}>/dev/stdout` to re-enable 
But this workaround works globally and may lead to breakage of other programs.
Maybe this can be done only for the `autocd` feature and thus not break other features.

@akinomyoga What is your opinion on this? Would you add this to ble.sh?

**Comment by akinomyoga** [comment]
The `autocd` feature is not a part of `ble.sh`'s feature, and `ble.sh` doesn't touch anything related to `autocd`. The `autocd` feature is handled by the command interpreter by Bash, so there is no way for a script to interfere with the behavior.

**Reply by Jai-JAP** [reply]
Yeah I am aware of that. 
I was just wondering that when the command is a dir name ble.sh can just redirect BASH_XTRACEFD TO null and then restore it after execution otherwise let it be as is if that's a regular command. 

This behaviour can be optionally disabled or enabled by an bleopt setting.

**Reply by akinomyoga** [reply]
I don't think the `BASH_XTRACEFD` hack is the right solution. At least, it's broken in many ways. It leaves the file descriptors kept open in child processes, and the background processes started by a session block closing the terminal window after closing the session, which causes a hanging terminal. `BASH_XTRACEFD` is the feature for `set -x` and not for `autocd`, so it conflicts with the feature `set -x`. I won't add such a broken feature, which requires users to understand how it works and in what situation it would be broken.

**Reply by akinomyoga** [reply]
> * Maybe this can be done only for the `autocd` feature and thus not break other features.

It is unpredictable whether the command line contains the command name that would be interpreted using `autocd` in general. It can be done *only when* the command line contains a single command name that doesn't match a command name but matches a directory name. I wouldn't like to add a feature that only works in a limited case and does not cover all the cases where `autocd` happens.

Even if there is a way to detect the general cases of `autocd`, there are still conflicts and problems raised in the above reply.

**Reply by akinomyoga** [reply]
Maybe I could instead add a magic expansion that converts the directory name `<dirname>` in the command context to `cd <dirname>`, so that the user only inputs the directory name, yet the executed command line would be `cd <dirname>`.

Or, another possibility would be to implement suffix aliases of `zsh` (`alias -s`, which defines the expansion of the filename with a particular filename extension in the command-name context), and to allow users to define a rule for the directory name as a special case.

**Reply by Jai-JAP** [reply]
> Maybe I could instead add a magic expansion that converts the directory name '<dirname>' in the command context to `cd <dirname>`, so that the user only inputs the directory name, yet the executed command line would be `cd <dirname>`.
> 

This would be a viable solution for this specific problem. The only need is that history should still reflect the command as `<dirname>` while executing `cd <dirname>`

> Or, another possibility would be to implement suffix aliases of `zsh` (`alias -s`, which defines the expansion of the filename with a particular filename extension in the command-name context), and to allow users to define a rule for the directory name as a special case.
> 
Well I don't know about that...

**Reply by akinomyoga** [reply]
> The only need is that history should still reflect the command as `<dirname>` while executing `cd <dirname>`

Why is it needed? I think if the command before the line editor's expansion is directly put in the history, it would cause more problems. The expansion of `<dirname>` into `cd <dirname>` is not Bash's feature but the feature of ble.sh as an editor.

Then, first, it is strange to put the intermediate state of the line editor in the history, which is not the command actually executed by Bash. If one later tries to use the history entry without ble.sh, it will not work as expected even though it is put in the command history of Bash.

Next, the magic expansion only happens with a certain condition, which is not preserved in the command history. If the directory name that was expanded by the magic expansion is directly put in the history, and the user picks the command from the history and executes it, it can be directly executed without the expansion and cause the message `cd -- <dirname>`, which you want to avoid.

**Reply by Jai-JAP** [reply]
Oh! Now I get it completely. That's right it may cause an issue for many users. It's better to just add `cd <dir>` to the history.

Can you add this feature to ble.sh?

**Reply by akinomyoga** [reply]
In the latest commit b6344b3be1978695889371de83ac4d15352e4fee, I added support for `autocd` in [`bleopt edit_magic_expand`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#user-content-bleopt-edit_magic_expand) and [`bleopt edit_magic_accept`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#user-content-bleopt-edit_magic_accept). Please check the descriptions there for details. For example, you can set the following:

```bash
# blerc

bleopt edit_magic_expand=history:sabbrev:autocd
bleopt edit_magic_accept=verify-syntax:autocd
```

**Reply by Jai-JAP** [reply]
That's nice. I am able to get it working the way I wanted. Thanks for your work.
